Add Client Commands

coehlrich opened the pull request:
Updated version of #6952 which didn't have much happening since October.

There are a couple changes:

- Renamed the event to `RegisterClientCommandsEvent`.
- When running a client command add the original message to the sent messages history.

GotoLink committed:
Allow command to be registered for client side execution

Leverage CommandDispatcher from Brigadier on client side.
Add CommandRegistryEvent and patch ClientPlayNetHandler to register
client commands to this dispatcher right after construction.
Patch ChatScreen and CommandSuggestionHelper (AT) to capture parsing
data before message is sent to ClientChatEvent and to server.
Patch Commands to remove empty lambdas of server commands from being
sent to client (and thus execute there when they should not).
Copy Commands method 'handleCommand' error handling process into
ForgeHooksClient.
Remove obsolete comment in Screen.
Add ClientCommandTest to illustrate registration, format and execution
of a client side command.

coehlrich committed:
Resolve conflicts with old client commands pr

coehlrich committed:
rename client commands register event

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
When executing a client command add it to the sent history

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

pupnewfster commented on src/main/java/net/minecraftforge/client/ForgeHooksClient.java:
```diff
@@ -61,10 +65,7 @@
-import net.minecraft.util.Hand;
-import net.minecraft.util.MovementInput;
-import net.minecraft.util.ResourceLocation;
+import net.minecraft.util.*;
```
Forge *tries* to stay away from wildcard imports (you have another being added lower down in this file as well)

coehlrich committed:
Remove the added * imports

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich commented:
I just tried suggestions on both a single player world and on an offline mode vanilla server run on the same PC and both times suggestions worked fine.

On single player world:
![image](https://user-images.githubusercontent.com/22903392/119755801-bbfbb680-bef6-11eb-8719-f94f4308a37a.png)

On offline mode vanilla server:
![image](https://user-images.githubusercontent.com/22903392/119755455-2eb86200-bef6-11eb-9776-f017922c341c.png)

kashike commented on patches/minecraft/net/minecraft/command/Commands.java.patch:
```diff
@@ -25,3 +25,11 @@
+                argumentbuilder.executes((p_197053_0_) -> {
+                   return 0;
+                });
++               argumentbuilder.executes(null);
```
Why?

sciwhiz12 replied:
See my previous lengthy explanation. As explained in the comment in the original PR, this  applies to the integrated server. The `SCommandListPacket` patch applies to the dedicated server (over the network syncing).

kashike commented on patches/minecraft/net/minecraft/network/play/server/SCommandListPacket.java.patch:
```diff
@@ -0,0 +1,10 @@
+                   this.field_197726_a.executes((p_197724_0_) -> {
+                      return 0;
+                   });
++                  this.field_197726_a.executes(null);
```
Why?

sciwhiz12 replied:
I'm not the author of the PR, but after reading the original PR and the justification given by the original PR's author (https://github.com/MinecraftForge/MinecraftForge/pull/6952#discussion_r458315650), I believe I understand the purpose of this.

**My suggestion**: As said in the linked review comment from the original PR, please add a comment briefly explaining why Forge sets the `executes` lambda of commands to be `null`.

***semi TL;DR: This allows the client hook to know that there is no client-side command, so it sends the command over to the server for final processing.***

### Background

This PR allows clients to register their custom commands to the client-side `CommandDispatcher<ISuggestionsProvider>` (as opposed to the server-side `CommandDispatcher<CommandSource>`). But, the commands are usually sent over to the server for final processing. 

### PR Logic

To allow the client-side custom commands to run, the PR manually executes the client-side dispatcher with the command. If there is no valid command found, then it delegates to the server for final execution. This happens when the parsed command tree does not end in a valid `executes` lambda. When no valid command is found, a `CommandSyntaxException` with the [`dispatcherUnknownCommand()`](https://github.com/Mojang/brigadier/blob/cf754c4ef654160dca946889c11941634c5db3d5/src/main/java/com/mojang/brigadier/exceptions/BuiltInExceptionProvider.java#L59) type is thrown. 

The hook captures command syntax exceptions, and if it captures one [with this specific type](https://github.com/coehlrich/MinecraftForge/blob/4b0d0484a6c6dbb5e293f2413b21909466d63ec8/src/main/java/net/minecraftforge/client/ForgeHooksClient.java#L854-L857), it assumes that there is no client-side command (otherwise execution would have succeeded, or an exception without that specific type is thrown), and thus throws it over to the server for processing.

(See [`ForgeHooksClient#sendMessage`](https://github.com/coehlrich/MinecraftForge/blob/4b0d0484a6c6dbb5e293f2413b21909466d63ec8/src/main/java/net/minecraftforge/client/ForgeHooksClient.java#L841) for the logic of the hook.)

### Modifying Vanilla's Behavior

In the original vanilla logic, when the client receives the server-side command information to construct a replica of the command tree (stored in the `CommandDispatcher<ISuggestionsProvider>`), it adds in empty lambdas for the command trees. This means that the explained logic would never work: since the command is valid due to the existence of the `executes` lambda, it gets executed _on the client_ (as the logic assumes if a command executes, it must be a custom client-side command) and it is never sent to the server for actual execution.

To allow the logic of the PR to work, this patch sets the `executes` parts of the replica command tree to be `null` instead of an empty (executing) lambda. This causes the command dispatcher to throw a command syntax exception with the aforementioned type, which allows the hook's logic to intercept it and send the command over to the server.

### Alternative

Possibly, this PR could create its own command syntax exception type, and instead of `null`-ing the lambdas, throws the exception with that custom type, and the hook would recognize that type specifically as the marker to send the command over to the server for real processing.

This has the added benefit of only sending valid server-listed commands (as defined in the replica command tree) over to the server for processing, while swallowing and presenting to the user if there was a genuine error with the client-side custom command (such as accidentally forgetting to add an `executes`).

coehlrich committed:
Re-add comments from previous PR while also mentioning that setting the
execute field to null while reading the packet is also for non-forge
servers instead of only vanilla servers and add more comment to the
sendMessage method for why the execute field needs to be set to null

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Made only valid server side commands get sent to the server and add a
CommandSource for the client

I want to change the place that the executes field is set to send to the
server since the lambda contains a client side class with the lambda
being in 2 classes that is used on a dedicated server (but currently it
still runs on a dedicated server fine)

Another thing I want to change is how the ClientCommandSource is used
since currently to use it in a command you need to cast the
CommandContext twice to use the ClientCommandSource for vanilla
arguments along with making sure that the modder knows that the getLevel
and getServer methods won't actually do anything or throw an exception
on the client

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Forgot to remove this line

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Use a seperate CommandDispatcher for client commands

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich commented:
I have separated the suggestions commands dispatcher from the client commands dispatcher (which allowed for removing the executes call when reading and writing packets), implemented a different form of the alternative that sciwhiz mentioned where instead of it throwing a custom exception when a server command get's executed the command is instead set to send the message to the server, I also created a ClientCommandSource with which most argument types that use a command source can be used apart from argument types that also use the world (since the return type of the method that the vanilla command source has is the server world) or the server, I also added a comment to the sendMessage method which was written prior to the suggestions commands dispatcher and the client commands dispatcher being split.

coehlrich committed:
Re-add support for suggestions

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Made the suggestions converter a function

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Stopped client commands from asking the server for suggestions

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

coehlrich committed:
Made the mergeServerCommands method use the vanilla suggestion provider
when copying server commands and again when copying the suggestions
dispatcher to the client commands dispatcher and changed to using the
getSource from the last commit for the copying client commands to the
suggestions and add a / when sending a server command to the server if
it doesn't already start with a / and remove the getSuggestions method
that I forgot to remove in a previous commit

Signed-off-by: coehlrich <coehlrich@users.noreply.github.com>

TheCurle commented:
This is a beast of a PR that none of us really feel comfortable triaging. 

We'll defer this until someone that knows what they're doing (like fry, preferably) takes a look at it.

This may just be something that has to be integrated upstream...

coehlrich committed:
Merge remote-tracking branch 'upstream/1.16.x' into 1.16.x

coehlrich committed:
Merge remote-tracking branch 'upstream/1.17.x' into
1.17.x-clientcommands

coehlrich committed:
Merge remote-tracking branch 'upstream/1.17.x' into 1.17.x-clientcommands

coehlrich committed:
Forgot to do genPatches

coehlrich commented:
Updated to the 1.17.x branch and the failed build's seems to be because of Jenkin's and not this PR.

coehlrich committed:
Merge remote-tracking branch 'upstream/1.17.x' into 1.16.x

coehlrich committed:
Add/Update documentation and change ClientCommandSource parameter names

coehlrich commented:
I want to override the `getServer` and `getLevel` methods in `ClientCommandSource` so that they throw an exception when trying to run in a client command but I'm not currently sure which exception to throw. I also want to add a separate method that can be used to get the un-sided Level.

coehlrich committed:
Changes:

Throw UnsupportedOperationException when attempting to use the
getServer/getLevel methods in a client command

Do printStackTrace when an generic exception is thrown during a client
command

Move the client commands test mod to the end of mods.toml

Made vanilla argument's that used getServer/getLevel where the things
they need exist on both sides able to be used on the client

coehlrich commented:
I have made a few more changes

Some things about the last change I mentioned in the commit message:

- The getWritableObjective for ObjectiveArgument will always be the same as getObjective when no objective's have been added client side since the server seems to always send objectives with the dummy type.
- The check that BlockPosArgument does to check if a block pos is in a loaded chunk always returns true on the client but the in world bounds check still works.
- Just remembered about the predicate argument's but I think I will wait for a review first.

coehlrich committed:
Merge remote-tracking branch 'upstream/1.17.x' into 1.16.x

Conflicts:
	src/test/resources/META-INF/mods.toml

coehlrich committed:
Merge remote-tracking branch 'upstream/1.17.x' into 1.16.x

Conflicts:
	src/test/resources/META-INF/mods.toml

coehlrich committed:
Merge remote-tracking branch 'upstream/1.18.x' into 1.16.x

Conflicts:
	patches/minecraft/net/minecraft/client/multiplayer/ClientPacketListener.java.patch

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandSource.java:
```diff
@@ -0,0 +1,126 @@
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class ClientCommandSource extends CommandSourceStack
```
This should be named `ClientCommandSourceStack`, to follow the naming scheme of its superclass and to clarify it is different from `CommandSource`. Additionally, javadocs should be created for this class and each of its methods (along with `IForgeCommandSourceStack`).

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+                }, (suggestions) -> null);
+    }
+
+    private static <S, T> void mergeCommandNode(CommandNode<S> sourceNode, CommandNode<T> resultNode, Map<CommandNode<S>, CommandNode<T>> sourceToResult,
+            S canUse, Command<T> execute, Function<SuggestionProvider<S>, SuggestionProvider<T>> sourceToResultSuggestion)
```
This and the `toResult` methods should have informative javadocs describing what they do and their parameters, for easier maintenance in the future.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+            ArgumentBuilder<T, ?> resultBuilder = null;
+            if (sourceNode instanceof ArgumentCommandNode<?, ?>)
+            {
+                ArgumentCommandNode<S, ?> sourceArgument = (ArgumentCommandNode<S, ?>) sourceNode;
```
This could possibly use pattern matching `instanceof` -- my IDE reports that `ArgumentCommandNode<S, ?> sourceArgument` doesn't cause an error, but I suggest a bit more testing to test and confirm it.

coehlrich replied:
The first type parameter for `ArgumentCommandNode` is used for the type parameter for `CommandNode` so the cast should be safe and likewise for `LiteralCommandNode` . Pattern matching `instanceof` can't have type parameters while this requires the type parameter.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+            }
+            else if (sourceNode instanceof LiteralCommandNode<?>)
+            {
+                LiteralCommandNode<S> sourceLiteral = (LiteralCommandNode<S>) sourceNode;
```
This could possibly use pattern matching `instanceof` -- my IDE reports that `LiteralCommandNode<S> sourceLiteral` doesn't cause an error, but I suggest a bit more testing to test and confirm it.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+                LiteralCommandNode<S> sourceLiteral = (LiteralCommandNode<S>) sourceNode;
+                resultBuilder = LiteralArgumentBuilder.literal(sourceLiteral.getLiteral());
+            }
+
```
What happens if a node is neither a `LiteralCommandNode` or an `ArgumentCommandNode`? Currently, this will cause a `NullPointerException`, but I'd like for the behavior to be defined explicitly.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+    private static <S, T> CommandNode<T> toResult(CommandNode<S> sourceNode, Map<CommandNode<S>, CommandNode<T>> sourceToResult, S canUse, Command<T> execute,
+            Function<SuggestionProvider<S>, SuggestionProvider<T>> sourceToResultSuggestion)
+    {
+        if (!sourceToResult.containsKey(sourceNode))
```
You could avoid a level of indentation by inverting this check and returning early the result of `sourceToResult.get(sourceNode)` instead.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+     * 
+     * @param currentParse
+     *            current state of the parser for the message
+     * @return false leaves the message to be sent to the server, true means it should be caught before {@link net.minecraft.client.gui.screen.ChatScreen#func_231161_c_(String)}
```
These javadocs need to be updated for the new names.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+            Component component = new TranslatableComponent("command.failed");
+            component.getStyle().withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, message));
+            Minecraft.getInstance().player.sendMessage(new TextComponent("").append(component).withStyle(ChatFormatting.RED), Util.NIL_UUID);
+            generic.printStackTrace();
```
This should be logged with a message instead of printed into `System.out`.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/chat/ClientCommandTest.java:
```diff
@@ -0,0 +1,120 @@
+                        .then(Commands.literal("registeredsuggest").then(
+                                Commands.argument("block", ResourceLocationArgument.id())
+                                        .suggests(SuggestionProviders.AVAILABLE_BIOMES)
+                                        .executes(new TestCommand())))
```
Instead of creating an object, this could a method reference to a static method instead.

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/chat/ClientCommandTest.java:
```diff
@@ -0,0 +1,120 @@
+    private void init(RegisterClientCommandsEvent event)
+    {
+        event.getDispatcher().register(
+                Commands.literal("clientcommandtest")
```
Each of these subcommands should have a small comment explaining what they are testing for.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+        commands = new CommandDispatcher<>();
+        MinecraftForge.EVENT_BUS.post(new RegisterClientCommandsEvent(commands));
+        RootCommandNode<SharedSuggestionProvider> serverCommandsCopy = new RootCommandNode<>();
+        mergeCommandNode(serverCommands, serverCommandsCopy, new HashMap<>(), Minecraft.getInstance().getConnection().getSuggestionsProvider(),
```
These three invocations of `mergeCommandNode` should be documented with comments explaining what they do and why they do so.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,226 @@
+
+public class ClientCommandHandler
+{
+    private static CommandDispatcher<CommandSourceStack> commands = null;
```
Should the client-side command dispatcher be exposed too, like how `Commands#getDispatcher()` exposes the server-side command dispatcher?

sciwhiz12 reviewed (changes requested):
First off, please update this PR to 1.18.1. 🚀
Second, I've done a bit of testing for this PR, using the test client and both a (forgedev) Forge server and a vanilla server, and it seems to work fine without problems. However, aside from my review comments and a general call to add more javadocs, I'd like a few points to be addressed (with test mods to confirm the behavior when possible):

- How does this handle command requirements, via `requires`?
- How does this handle forking commands (`ArgumentBuilder#isFork()`), or those with custom redirect modifiers (`ArgumentBuilder#getRedirectModifier()`)?
- There should be documentation somewhere on the behaviors of at least `ResourceLocationArgument.getAdvancement`, `ResourceLocationArgument.getRecipe`, and `ObjectiveArgument.getObjective` when called from a client command. Specifically, from my testing, they can only return those objectives, recipes, and advancements visible to the player -- advancements shown on the advancements screen, recipes shown in the recipe book, and objectives set to a display location (`/scoreboard objectives setdisplay`) that are shown to the player.

I encourage people to try out this PR locally and make their own client commands to see if there are any changes or fixes which needs to be made to this PR. 

I'm personally looking to have this merged in soon:tm:, since this is a good feature that's been requested for a good amount of time and it would be a shame to have this keep sitting here without any progress. For the record, I am willing to take over maintenance of this feature once this gets into Forge.

coehlrich committed:
Merge remote-tracking branch 'upstream/1.18.x' into 1.16.x

Conflicts:
	src/test/resources/META-INF/mods.toml

coehlrich committed:
Added suggestions from review comments

coehlrich committed:
Add test for command requirements

coehlrich committed:
Add test for forks and made forks from client side only work for client
side with similar for server side

coehlrich committed:
add javadocs to source stack classes

coehlrich committed:
Mention 2 argument types that behave differently when used on the client

All recipes get sent to the client even when they aren't visible to the
player, they get added to the client recipe book when they are visible

coehlrich commented:
Recipes get sent to the client and stored in the RecipeManager even when they aren't visible in the recipe book. The recipe book instead uses `ClientRecipeBook` for recipes to show.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+     * {@link net.minecraft.commands.Commands#performCommand(CommandSourceStack, String)} for reference
+     * 
+     * @param currentParse
+     *            current state of the parser for the message
```
This needs to match the parameter name.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+     * 
+     * @param currentParse
+     *            current state of the parser for the message
+     * @return false leaves the message to be sent to the server, true means it should be caught before {@link net.minecraft.client.gui.screen.ChatScreen#func_231161_c_(String)}
```
The reference to the `ChatScreen` method needs to be fixed.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+        // Copies the server side commands into a temporary server side commands root node to be used later without the client commands
+        RootCommandNode<SharedSuggestionProvider> serverCommandsCopy = new RootCommandNode<>();
+        mergeCommandNode(newServerCommands.getRoot(), serverCommandsCopy, new IdentityHashMap<>(),
+                Minecraft.getInstance().getConnection().getSuggestionsProvider(), (suggestions) -> 0, (suggestions) -> null);
```
Lambda parameters should be better named:
```suggestion
                Minecraft.getInstance().getConnection().getSuggestionsProvider(), (context) -> 0, (suggestions) -> null);
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+        // Copies the server side commands into the client side commands so that they can be sent to the server as a chat message
+        mergeCommandNode(serverCommandsCopy, commands.getRoot(), new IdentityHashMap<>(), Minecraft.getInstance().getConnection().getSuggestionsProvider(),
+                (source) -> {
+                    Minecraft.getInstance().player.chat((source.getInput().startsWith("/") ? "" : "/") + source.getInput());
```
```suggestion
                (context) -> {
                    Minecraft.getInstance().player.chat((context.getInput().startsWith("/") ? "" : "/") + context.getInput());
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+        // Copies the client side commands into the server side commands to be used for suggestions
+        mergeCommandNode(commands.getRoot(), newServerCommands.getRoot(), new IdentityHashMap<>(), getSource(), (suggestions) -> 0, (client) -> {
+            SuggestionProvider<SharedSuggestionProvider> suggestionProvider = SuggestionProviders
+                    .safelySwap((SuggestionProvider<SharedSuggestionProvider>) (SuggestionProvider<?>) client);
```
```suggestion
        mergeCommandNode(commands.getRoot(), newServerCommands.getRoot(), new IdentityHashMap<>(), getSource(), (context) -> 0, (suggestions) -> {
            SuggestionProvider<SharedSuggestionProvider> suggestionProvider = SuggestionProviders
                    .safelySwap((SuggestionProvider<SharedSuggestionProvider>) (SuggestionProvider<?>) suggestions);
```

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+            Component component = new TranslatableComponent("command.failed");
+            component.getStyle().withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, message));
+            Minecraft.getInstance().player.sendMessage(new TextComponent("").append(component).withStyle(ChatFormatting.RED), Util.NIL_UUID);
+            LOGGER.error("Error executing client command", generic);
```
This logging message should include the failing command.

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+        if (sourceToResult.containsKey(sourceNode))
+            return sourceToResult.get(sourceNode);
+
+        ArgumentBuilder<T, ?> resultBuilder = null;
```
The initializer to `null` is unneeded, as variables for reference objects are by default set to `null` and this variable will always have a value (or an exception is thrown).

sciwhiz12 reviewed (changes requested):
A few more small changes, and I think this looks fine to me.

coehlrich committed:
Add suggestions and update javadoc for the sendMessage method

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+        {
+            TextComponent message = new TextComponent(generic.getMessage() == null ? generic.getClass().getName() : generic.getMessage());
+            Component component = new TranslatableComponent("command.failed");
+            component.getStyle().withHoverEvent(new HoverEvent(HoverEvent.Action.SHOW_TEXT, message));
```
This won't work, as `Style#withHoverEvent` returns a new `Style` (as they are immutable objects).

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+            {
+                int position = Math.min(syntax.getInput().length(), syntax.getCursor());
+                MutableComponent details = new TextComponent("").withStyle(ChatFormatting.GRAY);
+                details.getStyle().withClickEvent(new ClickEvent(ClickEvent.Action.SUGGEST_COMMAND, reader.getString()));
```
This won't work, as `Style#withClickEvent` returns a new `Style` (as they are immutable objects).

sciwhiz12 commented on src/main/java/net/minecraftforge/client/ClientCommandHandler.java:
```diff
@@ -0,0 +1,312 @@
+     */
+    public static boolean sendMessage(String sendMessage)
+    {
+        ClientCommandSourceStack source = getSource();
```
A small nitpick: This should be moved to below the `if`-check, so the source isn't constructed if not needed.

sciwhiz12 reviewed (changes requested):
A bit more changes that _almost_ got by me. 😅

coehlrich committed:
Add suggested changes

sciwhiz12 reviewed (approved):
:shipit:

coehlrich committed:
Merge remote-tracking branch 'upstream/1.18.x' into 1.16.x

coehlrich committed:
Merge remote-tracking branch 'upstream/1.18.x' into 1.16.x

Conflicts:
	src/test/resources/META-INF/mods.toml

coehlrich committed:
Merge remote-tracking branch 'upstream/1.18.x' into 1.16.x

Conflicts:
	src/test/resources/META-INF/mods.toml
